src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -352,11 +352,15 @@ internal NamespaceOrTypeOrAliasSymbolWithAnnotations BindNamespaceOrTypeOrAliasS
                             {
                                 diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(Compilation, NonNullTypesContext, typeArgument), nullableSyntax.QuestionToken.GetLocation());
                             }
-                            else if (!typeArgument.IsValueType)
+                            else if (!typeArgument.IsValueType && !typeArgument.IsErrorType())
                             {
                                 Symbol.ReportNullableReferenceTypesIfNeeded(Compilation, NonNullTypesContext, diagnostics, nullableSyntax.QuestionToken.GetLocation());
                             }
                         }
+                        else
+                        {
+                            diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(Compilation, NonNullTypesContext, typeArgument), nullableSyntax.QuestionToken.GetLocation());
```

Comment:
>diagnostics.Add(new LazyMissingNonNullTypesContextDiagnosticInfo(Compilation, NonNullTypesContext, typeArgument), nullableSyntax.QuestionToken.GetLocation()); [](start = 28, length = 158)

I need some help to understand this. 
Previously, the check for misused `?` was checked by `ReportNullableReferenceTypesIfNeeded` in the various symbols. That minimized the use of lazy diagnostics.
What's the scenario that makes using a lot more lazy diagnostics necessary?

Also, the comment above should be updated, since reflects previous design.

Reply:
> Previously, the check for misused ? was checked by ReportNullableReferenceTypesIfNeeded in the various symbols. That minimized the use of lazy diagnostics.

Diagnostics should be reported when a type is bound. Old approach had obvious holes (some types were not checked, can be observed in test) and was fragile - every one implementing a new symbol was supposed to worry about not missing the checks and figure out the right location, etc. In addition, an attempt to report this diagnostics eagerly in ConstraintHelper.cs was causing a cycle in some scenarios.
I will remove the "// Types created outside executable context should be checked by the responsible symbol (the method symbol checks its return type, for instance)." comment.

---
In reply to: [217135590](https://github.com/dotnet/roslyn/pull/29809#discussion_r217135590) [](ancestors = 217135590)

Reply:
Regarding the question of fragility:
We used to pair calls to `ReportNullableReferenceTypesIfNeeded` and `ReportAnnotatedUnconstrainedTypeParameters` with calls to `EnsureNullableAttributeExists`.
If we were missing some cases of the former, then we're likely missing some cases of the latter too. Should we also move the call to `EnsureNullableAttributeExists` here? (file an issue?)

---
In reply to: [217141886](https://github.com/dotnet/roslyn/pull/29809#discussion_r217141886) [](ancestors = 217141886,217135590)

Reply:
> Should we also move the call to EnsureNullableAttributeExists here? (file an issue?)

No. That check is in the right place. Who synthesizes the attribute is responsible for ensuring it exists.

---
In reply to: [217159561](https://github.com/dotnet/roslyn/pull/29809#discussion_r217159561) [](ancestors = 217159561,217141886,217135590)

Reply:
Also note, that not every annotated type results in an attribute synthesized somewhere  

---
In reply to: [217160425](https://github.com/dotnet/roslyn/pull/29809#discussion_r217160425) [](ancestors = 217160425,217159561,217141886,217135590)